### ISSUE TITLE
DriveTrain - velocityDrive taking joysticks as a parameter

### ISSUE THREAD

**Issue by bjmcternan** [Team293/InfiniteRecharge#3]
@Team293/2020:
In src/main/java/frc/robot/subsystems/Drivetrain.java on lines 161 - 172 you have:

`public void velocityDrive(Joystick left, Joystick right){`
`        double leftPos = left.getY();`
`        double rightPos = right.getY();`
`        double retval = 0.0;`
`        boolean useSlowModifier = false;`
` `
`       if(Robot.oi.rightJoy.getTrigger()){`
`            useSlowModifier = true;`
`        }`
`        if(Robot.oi.leftJoy.getTrigger()){`
`            useSlowModifier = true;`
`        }`

This isn't correct. `velocityDrive` should not be passed Joysticks as a parameter. You can pass in the getTrigger() stuff to `velocityDrive` as a third `boolean` parameter.

Change it to `velocityDrive(double leftPos, double rightPos, boolean useSlowModifier)` on line 161 and fix the rest of the code below it. Anyone that calls `velocityDrive` can get the position of the joysticks and then get the status of the triggers and pass that down to the Drivetrain susbsystem through the velocityDrive function.

**Comment by bjmcternan** [comment]
By passing in `double leftPos` and `double rightPos` the autonomous trajectory code can use this function to move. As it is right now the autonomous trajectory stuff doesn't have a `Joystick` to use to request drivetrain velocity changes

**Comment by bjmcternan** [comment]
@ReyRiordan and @kailyl can you both fix the drivetrain stuff today? The changes made in Show it are good and should be copied over

**Comment by cartha353** [comment]
I believe they won't be here today but Jonah and I are free as we have
nothing to do, the mechanical team is still working on a lot of things and
the coding team is far ahead. Jonah and I are also planning to add a button
in the drivetrain branch to reverse the motors so the driver can easily
switch from pickup to shooting.

**Comment by bjmcternan** [comment]
Just so long as someone fixes the drivetrain before we forget and this
issue gets into the rest of the code. A backup button sounds interesting.
Is the shooter backwards from the drivetrain?

**Comment by cartha353** [comment]
Yeah the pickup and the shoot are in different sides of the robot.

**Comment by BobTheLegend** [comment]
Ok the issue should be resolved with the new commit to the Drivetrain branch. Sorry it took so long, the people who fixed it originally didn't commit or push.

**Comment by bjmcternan** [comment]
Almost. See my comment in Drivetrain.java. lines 163-168 need to be removed

**Comment by bjmcternan** [comment]
Also, the issue isn't complete until this change is merged down to the master branch and the other branches are updated!

**Comment by cartha353** [comment]
We removed the code and we are not going to merge it into master until we
test it.

**Comment by bjmcternan** [comment]
Good idea! Issues shouldn't be closed until tested 😁

**Comment by mckillip** [comment]
Hi Brennan:

Thanks as always for keeping your hand on the tiller with the coding 
folks!

As for the direction control button ... since the ball pickup and the 
shooter are on opposite sides, it begs the question which is the "front" 
of the robot.  Initial start of competition will have our robot on the 
far side of the field, with the ball pickup facing the driver station 
and the shooter facing the goal.  Once autonomous is done, teleop mode 
would have Jimmy pushing the sticks to have the robot back up to the 
driver station if the ball pickup were the "front" - a rather awkward 
feature.  While this is trainable with sufficient practice, in 2016 we 
had a mode select switch that let the joysticks change "sense", with LED 
strips on both sides of the robot that would change color based upon 
which direction was forward (white lights = "headlights", red = 
"taillights").  Noah Hillman, Mark's son, liked that operation and said 
it was easy to drive.  Mark suggested the coders include that feature so 
Jimmy can try it and decide what he wants when the builders turn the 
drivetrain over to the coders and, ultimately, Jimmy for practice.

That's the "back story"!

Ed thinks the robot should be drivable close to Wednesday; some 
preliminary checkout on the electronics was being done when we called it 
a day today.

**Comment by bjmcternan** [comment]
Thanks Bob! I agree, a "reverse" button seems useful in this configuration.
I think the dev team is also working towards adding that as a feature now.
The team has done a good job fixing some bugs I found and handling some
nit-picky things.

**Comment by mckillip** [comment]
Hi Programmers:

I was talking with Jonah on Sunday and said the documentation on how the added WPILib features allow for path following seemed to be a “black box” that scared me, as there didn’t seem to be a way to understand it sufficiently to tweak it in the event that one didn’t get the results you wanted.  I did find a reference that may provide a clean feedback approach for tracking a trajectory (field x,y position and pointing direction, collectively called “pose” of the robot), and will research that some more, but had an interesting discussion with Jeff Keller (mentor at Nemesis 2590), a co-worker here at CDI.

Jeff told me that his team had “lost” the internal path following knowledge they had when a key programmer graduated, who had a rather awkward approach toward tweaking the path and then the feedbacks needed to follow that path.  So starting from a blank sheet, the current team programmers merely walked through the current documentation, and followed the examples - and it worked!  He feels a key was the “characterization” of their robot, which is done using a python routine that runs on the driver station and collects the required data to help design the feedback gains.  We can do this, even with our internal velocity-feedback/feedforward system to check its performance.  The code development for the robot includes a “odometry” class that keeps tabs on the encoders on each side to help track the desired trajectory, and he showed me a video of their robot making a smooth S-turn from a dead stop.

What this means is I think the programmers, under Jonah’s guidance, should look at the documentation and walk through the process for implementing the trajectory following scheme there, and I will continue to entertain myself with an alternate feedback-driven approach.  That way we have "two lines in the water” and hopefully one of them will catch a fish.

Thanks for tolerating my decision changes, but this was useful new data from Jeff.  He also sent me some references on the algorithm - a slide deck from the 254 presentation (the YouTube video is at: https://youtu.be/8319J1BEHwM  which is a hour long!).  It has a little physics and suggests some calculus, but gives some insight on what is under the hood - which is essential to playing with a “black box”.

-B.
Dr. Robert McKillip, Senior Associate
Continuum Dynamics, Inc.
34 Lexington Ave, Ewing, NJ 08618
609-538-0444 x112
bob@continuum-dynamics.com

**Comment by BobTheLegend** [comment]
Ok, so the final drivetrain code has been merged into master and the issue with drivetrain code that originally started this has been resolved. As the code has been fixed and tested I'm now going to close this issue.

**Comment by bjmcternan** [comment]
@BobTheLegend stop closing the ticket when you didn't open it. I still think you have suspect code

**Comment by bjmcternan** [comment]
@BobTheLegend stop closing a ticket that you didn't open. I still think you have suspect code. I'll close it when I believe it is fixed

**Comment by bjmcternan** [comment]
This is really close. I think once the reversing is handled with a boolean I will close this

**Comment by bjmcternan** [comment]
Marking this fixed as of commit: https://github.com/Team293/InfiniteRecharge/commit/f7df80c9560734ef94f2fa2b05b10f42ccd290f4
